common/djangoapps/entitlements/utils.py

```diff
@@ -50,8 +55,11 @@ def is_course_run_entitlement_fulfillable(course_run_key, entitlement, compare_d
         and (not enrollment_end or enrollment_end > compare_date)
     )
 
+    # Is the user already enrolled in the Course Run
+    is_enrolled = CourseEnrollment.is_enrolled(entitlement.user, course_run_key)
+
     # Ensure the course run is upgradeable and the mode matches the entitlement's mode
     unexpired_paid_modes = [mode.slug for mode in CourseMode.paid_modes_for_course(course_run_key)]
     can_upgrade = unexpired_paid_modes and entitlement.mode in unexpired_paid_modes
 
-    return is_running and can_upgrade and can_enroll
+    return is_running and can_upgrade and (is_enrolled or can_enroll)
```

Comment:
So this doesn't check for the case that the learner is enrolled, the enrollment window is closed, but the upgrade window is still open. You'll need to add in that logic in perhaps a `can_verify` boolean, as this is checking only whether they're enrolled, and then always allow them to upgrade.

Reply:
can_upgrade would be False so you would get 
True AND False AND (TRUE or FALSE) which would = FALSE 

That would cover this case I believe.

Reply:
Completely missed the can_upgrade part, ignore this 🙃